File: runtime/micropython/night.py
"""night: the night policy, pure. docs/camera_operation.md §6, mirrors
dc_night_step / dc_night_predict in runtime/espidf/components/
dusty_core/src/night.c exactly (semantics), generalised so the long "sleep
toward dawn" figure can be recomputed on every probe, not only at the
moment night is entered -- this lets a wake resume mid-night from a saved
state and still land near the learned dawn.

State dict: {'night': 0/1, 'dark_n': int, 'start': epoch_s, 'hist': [..<=7]}
(newest length last). `night_step` never sleeps or does I/O; the caller
saves the state and deep-sleeps for `sleep_s`.
"""
import json

NIGHT_HIST = 7


def night_new():
    return {'night': 0, 'dark_n': 0, 'start': 0, 'hist': []}


def night_predict(state):
    """Median of the newest up-to-3 of up to 7 lengths. 0 with no history."""
    hist = state.get('hist') or []
    if not hist:
        return 0
    last = hist[-3:]
    s = sorted(last)
    k = len(s)
    if k % 2 == 1:
        return s[k // 2]
    return (s[k // 2 - 1] + s[k // 2]) // 2


def _night_sleep(state, cfg, now):
    if not state.get('hist'):
        return cfg['night_probe_s']
    predicted = night_predict(state)
    remaining = predicted - cfg['night_margin_s'] - (now - state.get('start', now))
    if remaining > cfg['night_probe_s']:
        return remaining
    return cfg['night_probe_s']


def night_step(state, lum, now, cfg):
    """(sleep_s, event). event in 'none','enter','probe','exit'. sleep_s is
    None only when not applicable (never, in practice: not-night returns
    cfg['period_s'])."""
    if not state.get('night'):
        if lum < cfg['lum_night']:
            state['dark_n'] = state.get('dark_n', 0) + 1
        else:
            state['dark_n'] = 0
        if state['dark_n'] >= cfg['night_confirm_n']:
            state['night'] = 1
            state['start'] = now
            state['dark_n'] = 0
            sleep_s = _night_sleep(state, cfg, now)
            return sleep_s, 'enter'
        return cfg['period_s'], 'none'

    # already in night: this wake is a probe
    if lum >= cfg['lum_day']:
        length = now - state.get('start', now)
        hist = state.get('hist') or []
        hist.append(length)
        state['hist'] = hist[-NIGHT_HIST:]
        state['last_len'] = length
        state['night'] = 0
        state['dark_n'] = 0
        return cfg['period_s'], 'exit'

    sleep_s = _night_sleep(state, cfg, now)
    return sleep_s, 'probe'


def night_load(path):
    st = night_new()
    try:
        with open(path) as f:
            d = json.loads(f.read())
        st['hist'] = list(d.get('hist', []))[-NIGHT_HIST:]
        st['night'] = int(d.get('night', 0))
        st['start'] = int(d.get('start', 0))
    except (OSError, ValueError):
        pass
    return st


def night_save(path, state):
    d = {'hist': list(state.get('hist', []))[-NIGHT_HIST:],
         'night': state.get('night', 0), 'start': state.get('start', 0)}
    try:
        with open(path, 'w') as f:
            f.write(json.dumps(d))
        return True
    except OSError:
        return False

File: runtime/espidf/components/dusty_core/src/night.c
/* night.c: night policy (camera_operation.md §6), pure function of state,
 * lum, now and cfg -- see dusty_core.h for exact field semantics. */
#include "dusty_core.h"
#include "json_lite.h"
#include "jsonw.h"

uint32_t dc_night_predict(const dc_night_t *st) {
    if (st->hist_n <= 0) return 0;
    int k = st->hist_n < 3 ? st->hist_n : 3;
    uint32_t vals[3];
    for (int i = 0; i < k; i++) {
        int idx = (st->hist_head - 1 - i + 2 * DC_NIGHT_HIST) % DC_NIGHT_HIST;
        vals[i] = st->hist[idx];
    }
    for (int i = 0; i < k; i++)
        for (int j = i + 1; j < k; j++)
            if (vals[j] < vals[i]) { uint32_t t = vals[i]; vals[i] = vals[j]; vals[j] = t; }
    if (k % 2 == 1) return vals[k / 2];
    return (vals[k / 2 - 1] + vals[k / 2]) / 2;
}

static void push_hist(dc_night_t *st, uint32_t len) {
    st->hist[st->hist_head] = len;
    st->hist_head = (st->hist_head + 1) % DC_NIGHT_HIST;
    if (st->hist_n < DC_NIGHT_HIST) st->hist_n++;
}

void dc_night_step(dc_night_t *st, int lum, uint32_t now_s, const dc_night_cfg_t *cfg, dc_night_out_t *out) {
    out->entered = 0;
    out->left = 0;
    out->night_len_s = 0;

    if (!st->in_night) {
        if (lum < cfg->lum_night) st->dark_n++;
        else st->dark_n = 0;

        if (st->dark_n >= cfg->night_confirm_n) {
            /* enter night */
            st->in_night = 1;
            st->night_start_s = now_s;
            st->dark_n = 0;
            st->slept_toward_dawn = 0;
            out->entered = 1;

            uint32_t sleep_s;
            if (st->hist_n > 0) {
                uint32_t predicted = dc_night_predict(st);
                uint32_t long_sleep = predicted > cfg->night_margin_s ? predicted - cfg->night_margin_s : 0;
                if (long_sleep > cfg->night_probe_s) {
                    sleep_s = long_sleep;
                    st->slept_toward_dawn = 1;
                } else {
                    sleep_s = cfg->night_probe_s;
                }
            } else {
                sleep_s = cfg->night_probe_s;
            }
            out->sleep_s = sleep_s;
        } else {
            out->sleep_s = cfg->period_s;
        }
        out->in_night = st->in_night;
        return;
    }

    /* already in night: this wake is a probe */
    if (lum > cfg->lum_day) {
        out->left = 1;
        out->night_len_s = now_s - st->night_start_s;
        push_hist(st, out->night_len_s);
        st->in_night = 0;
        st->dark_n = 0;
        st->slept_toward_dawn = 0;
        out->sleep_s = cfg->period_s;
    } else {
        out->sleep_s = cfg->night_probe_s;
    }
    out->in_night = st->in_night;
}

int dc_night_hist_to_json(const dc_night_t *st, char *buf, size_t n) {
    size_t off = 0;
    off = out_char(buf, off, n, '[');
    int oldest = (int)(((long)st->hist_head - (long)st->hist_n + DC_NIGHT_HIST) % DC_NIGHT_HIST);
    for (int i = 0; i < st->hist_n; i++) {
        int idx = (oldest + i) % DC_NIGHT_HIST;
        if (i > 0) off = out_char(buf, off, n, ',');
        off = out_fmt(buf, off, n, "%u", (unsigned)st->hist[idx]);
    }
    off = out_char(buf, off, n, ']');
    if (off >= n) return -1;
    buf[off] = '\0';
    return (int)off;
}

int dc_night_hist_from_json(dc_night_t *st, const char *json) {
    if (!json) return -1;
    const char *p = json;
    jl_skip_ws(&p);
    if (*p != '[') return -1;
    p++;
    jl_skip_ws(&p);

    uint32_t vals[DC_NIGHT_HIST];
    int count = 0;

    if (*p == ']') {
        p++;
    } else {
        for (;;) {
            jl_val_t v = {0};
            if (jl_parse_value(&p, &v) != 0 || v.type != JL_NUM) return -1;
            if (count < DC_NIGHT_HIST) vals[count] = (uint32_t)(v.num + 0.5);
            else { /* keep the newest DC_NIGHT_HIST: shift the window */
                for (int i = 1; i < DC_NIGHT_HIST; i++) vals[i - 1] = vals[i];
                vals[DC_NIGHT_HIST - 1] = (uint32_t)(v.num + 0.5);
            }
            count++;
            jl_skip_ws(&p);
            if (*p == ',') { p++; continue; }
            if (*p == ']') { p++; break; }
            return -1;
        }
    }

    int keep = count < DC_NIGHT_HIST ? count : DC_NIGHT_HIST;
    for (int i = 0; i < DC_NIGHT_HIST; i++) st->hist[i] = 0;
    for (int i = 0; i < keep; i++) st->hist[i] = vals[i];
    st->hist_n = keep;
    st->hist_head = keep % DC_NIGHT_HIST;
    return keep;
}
